File: src/ControlAlgorithm.cpp
#include "ControlAlgorithm.h"
#include <utility>

ControlAlgorithm::ControlAlgorithm(std::string name, std::vector<SignalPhase> phases)
    : name(std::move(name)), phases(std::move(phases)) {
    if (this->phases.empty()) {
        this->phases.emplace_back("all-red", 5, std::unordered_map<std::string, LightState>{});
    }
}

const std::string &ControlAlgorithm::getName() const { return name; }

void ControlAlgorithm::setMode(std::string modeName) { mode = std::move(modeName); }

const std::string &ControlAlgorithm::getMode() const { return mode; }

const SignalPhase &ControlAlgorithm::nextPhase() {
    const auto &phase = phases[currentPhase];
    currentPhase = (currentPhase + 1) % phases.size();
    return phase;
}

void ControlAlgorithm::reset() { currentPhase = 0; }

File: python/cross_roads_ai/control_algorithm.py
from typing import List
from .signal_phase import SignalPhase

class ControlAlgorithm:
    def __init__(self, name: str, phases: List[SignalPhase]):
        self._name = name
        self._phases = list(phases)
        self._index = 0
        self._mode = ""

    def nextPhase(self) -> SignalPhase:
        if not self._phases:
            raise RuntimeError("No phases available")
        phase = self._phases[self._index]
        self._index = (self._index + 1) % len(self._phases)
        return phase

    def getName(self) -> str:
        return self._name

    def setMode(self, mode: str) -> None:
        self._mode = mode

    def mode(self) -> str:
        return self._mode
